File: src/main/year2022/Day04.java
package year2022;

import template.Day;

import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class Day04 extends Day {

    public Day04() {
        super("input.txt", "04", "2022");
    }

    private static class Pair {
        int start;
        int end;

        public Pair(int start, int end) {
            this.start = start;
            this.end = end;
        }

        public boolean contains(Pair other) {
            return this.start <= other.start && other.end <= this.end;
        }

        public boolean overlaps(Pair other) {
            return this.start <= other.end && other.start <= this.end;
        }
    }

    private List<Pair> parseRanges(String line) {
        String[] rangeStrings = line.split(",");
        return Stream.of(rangeStrings)
                .map(rangeStr -> {
                    String[] bounds = rangeStr.split("-");
                    int start = Integer.parseInt(bounds[0]);
                    int end = Integer.parseInt(bounds[1]);
                    return new Pair(start, end);
                })
                .collect(Collectors.toList());
    }

    @Override
    public long part1() {
        List<String> inputList = getInput();
        List<List<Pair>> rangesList = inputList.stream()
                .map(this::parseRanges)
                .toList();

        return rangesList.stream()
                .filter(ranges -> {
                    Pair a = ranges.get(0);
                    Pair b = ranges.get(1);
                    return a.contains(b) || b.contains(a);
                })
                .count();
    }

    @Override
    public long part2(){
        List<String> inputList = getInput();
        List<List<Pair>> rangesList = inputList.stream()
                .map(this::parseRanges)
                .toList();

        return rangesList.stream()
                .filter(ranges -> {
                    Pair a = ranges.get(0);
                    Pair b = ranges.get(1);
                    return a.overlaps(b);
                })
                .count();
    }


}

File: src/main/year2022/day04/Day04.kt
package year2022.day04

import java.io.File

class Day04(text : String = "input.txt") {

    private val day = "04"
    private val input = File("src/resources/2022/$day/$text").readLines().asSequence().map { s ->
        s.split(",").map { r ->
        val (a , b) = r.split("-").map { it.toInt() }
        a..b
    } }
    private operator fun IntRange.contains(i: IntRange): Boolean = this.first <= i.first && this.last >= i.last
    fun part1(): Int = input.count{ (a, b) -> b in a || a in b }
    fun part2(): Int = input.count{ (a, b) -> a.first in b || b.first in a }
}
fun main() {
    val day = Day04()
    println(day.part1())
    println(day.part2())
}

